src/driver/compile.c

```diff
@@ -16,41 +16,206 @@
-    struct yf_args *
+static int yf_compile_project(struct yf_args *, struct yf_compilation_data *);
+static int yf_compile_files(struct yf_args *, struct yf_compilation_data *);
+static int yfc_analyse(
+    struct yf_compilation_data *,
+    struct yf_compile_analyse_job *
+);
+static int yfc_compile(
+    struct yf_compilation_data *,
+    struct yf_compile_compile_job *
 );
 static int yf_find_project_files(struct yf_project_compilation_data *);
 static int dump_tokens(struct yf_lexer *);
-static int yf_build_symtab(struct yf_file_compilation_data *);
+static int yf_build_symtab(struct yf_compile_analyse_job *);
 static int yf_validate_ast(
-    struct yf_project_compilation_data * pdata,
-    struct yf_file_compilation_data * fdata,
-    struct yf_args * args
+    struct yf_compilation_data * pdata,
+    struct yf_compile_analyse_job * adata
 );
 static int yf_do_cst_dump(struct yf_parse_node * tree);
-static int yf_cleanup(struct yf_project_compilation_data *);
+static int yf_cleanup(struct yf_compilation_data *);
+
+static inline const char * str_or_null(const char * s) {
+    return s ? s : "(null)";
+}
+
+static void yf_dump_compile_job(struct yf_compile_analyse_job * job, const char * label) {
+    fputs(label, YF_OUTPUT_STREAM);
+    switch (job->stage) {
+        case YF_COMPILE_LEXONLY:
+            fputs(" (DUMP_TOKENS) ", YF_OUTPUT_STREAM);
+            break;
+        case YF_COMPILE_PARSEONLY:
+            fputs(" (DUMP_CST) ", YF_OUTPUT_STREAM);
+            break;
+        case YF_COMPILE_ANALYSEONLY:
+            fputs(" (ANALYSE) ", YF_OUTPUT_STREAM);
+            break;
+        case YF_COMPILE_CODEGEN:
+            fputs(" (COMPILE) ", YF_OUTPUT_STREAM);
+            break;
+    }
+    if (!job->unit_info->parse_anew) {
+        fputs("(cached) ", YF_OUTPUT_STREAM);
+    }
+    fprintf(YF_OUTPUT_STREAM, "file=%s prefix=%s syms=%s out=%s\n",
+        job->unit_info->file_name,
+        str_or_null(job->unit_info->file_prefix),
+        str_or_null(job->unit_info->sym_file),
+        str_or_null(job->unit_info->output_file)
+    );
+}
+
+/**
+ * Create a list of jobs to be done to compile with the given arguments.
+ */
+static int yf_create_compilation_data(struct yf_args * args, struct yf_compilation_data * compilation) {
+    if (args->project) {
+        return yf_compile_project(args, compilation);
+    } else {
+        return yf_compile_files(args, compilation);
+    }
+}
 
 /**
  * This is it. This is the actual compile function for a set of arguments. It
  * just defers compilation to one of two functions, depending on whether
  * --project is enabled or not.
  */
 int yf_run_compiler(struct yf_args * args) {
-    
-    if (args->project) {
-        return yf_compile_project(args);
-    } else {
-        return yf_compile_files(args);
+
+    struct yf_compilation_data compilation;
+    struct yf_compilation_job * job;
+    int res = 0;
+
+    yf_create_compilation_data(args, &compilation);
+
+    if (res)
```

Comment:
Is this necessary? `res` has never been touched.

Reply:
It seems like it was supposed to be assigned from the call above, oops.